4 lab(custom controls): реализовал лабораторную 4 по предмету ООАиП.

RockInHead opened the pull request.

RockInHead committed:
feature(View): создал новый пользовательский элемент управления - ContactControl, переместил туда всю правую панель.

RockInHead committed:
feature(MainWindow.xaml): реализовал DataTemplate.

- DataTemplate отображает изображение по умолчанию, имя и номер телефона контакта.
- вернул отображение иконки и ограничение окна по размерам.

RockInHead committed:
feature(View): реализовал валидацию полей ввода.

- добавил в Contact.cs интерфейс IDataErrorInfo для передачии сообщения об ошибке и настройке условий валидации.
- создал свойство HasValidationErrors, указывающее, есть ли ошибки валидации. От него также стало зависеть свойство CanApplyContact.
- для сохранения оригинального, неотредактированного контакта создал поле _originalContact и в случае, если пользователь отменяет редактирование контакта, то  поля SelectedContact приравнивается к полям  _originalContact.

RockInHead committed:
feature(CotactControl): реализовал запрет ввода букв, пробелов и вставки из буфера для поля PhoneNumber.

RockInHead committed:
style(View): добавил xml-комменатриев, убрал лишнее.

sandra155464 commented on Programming/Contacts/Contacts/Controls/ContactControl.xaml.cs:
```diff
@@ -0,0 +1,63 @@
+        /// </summary>
+        /// <param name="sender">Элемент, вызвавший событие.</param>
+        /// <param name="e">Аргументы события, содержащие информацию о команде.</param>
+        private void TextBox_PreviewExecuted(object sender, ExecutedRoutedEventArgs e)
```
В методичке сказано, что нужно сделать так, чтобы в текстбокс с номером телефона нельзя было копировать только **недопустимые** символы из буфера обмена, а у тебя получается, что вообще ничего скопировать нельзя - надо реализовать проверку на допустимые символы и разрешать вставить из буфера обмена только их. Также запрет на копирование из буфера обмена вместо TextBox_PreviewExecuted можно реализовать с помощью TextBox_Pasting - в методичке тоже было сказано именно про него

sandra155464 commented on Programming/Contacts/Contacts/Model/Contact.cs:
```diff
@@ -1,9 +1,10 @@
+public class Contact : INotifyPropertyChanged, IDataErrorInfo, ICloneable
 {
     /// <summary>
     /// Поле, задающее имя контакта.
```
"Хранящее" вместо "задающее" - по аналогии с комментариями к другим полям

sandra155464 commented on Programming/Contacts/Contacts/Model/Contact.cs:
```diff
@@ -97,8 +101,25 @@ public string Email
+    /// <returns>Новый объект Contact с такими же значениями свойств.</returns>
+    public object Clone()
+    {
+        return new Contact
```
Можно переписать так:
``` cs
public object Clone() => new Contact(this.Name, this.PhoneNumber, this.Email);
```

sandra155464 commented on Programming/Contacts/Contacts/Model/Contact.cs:
```diff
@@ -108,5 +129,49 @@ protected void OnPropertyChanged(string propertyName)
+        {
+            switch (columnName)
+            {
+                case "Name":
```
Неправильно оформлены блоки case - правильнее писать так:
``` cs
case "Name":
{
// логика
}
```

sandra155464 commented on Programming/Contacts/Contacts/Model/Contact.cs:
```diff
@@ -108,5 +129,49 @@ protected void OnPropertyChanged(string propertyName)
+                        }
+                        break;
+                    }
+            }
```
Не хватает пустой строки после закрывающей фигурной скобки

sandra155464 commented on Programming/Contacts/Contacts/Model/Contact.cs:
```diff
@@ -108,5 +129,49 @@ protected void OnPropertyChanged(string propertyName)
+                            return "Номер телефона может содержать только цифры и символы '+()-'.";
+                        }
+                        break;
+                    }
```
Не хватает пустой строки после закрывающей фигурной скобки

sandra155464 commented on Programming/Contacts/Contacts/Model/Contact.cs:
```diff
@@ -108,5 +129,49 @@ protected void OnPropertyChanged(string propertyName)
+                        break;
+                    }
+            }
+            return null;
```
Тоже нужна пустая строка перед ```return null;```

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -78,6 +78,11 @@ public bool IsReadOnlyMode
+    /// <summary>
+    /// Возвращает или задает контакт до редактирования.
+    /// </summary>
+    private Contact OriginalContact { get; set; }
```
Если свойство закрытое, то лучше сделать вместо него поле

sandra155464 commented on Programming/Contacts/Contacts/Model/Contact.cs:
```diff
@@ -97,8 +101,25 @@ public string Email
+    /// Возвращает общую ошибку валидации объекта. 
+    /// В данном случае всегда null.
+    /// </summary>
+    public string Error => null;
```
Для чего это свойство? Оно нигде не используется

sandra155464 commented:
Оставила замечания по лабораторной

RockInHead committed:
refactoring(View): поправил оформление кода и комментариев.

- упростил реализацию метода Clone() в Contact.cs

RockInHead committed:
fix(ContactControl.xaml): исправил функцию вставки в текстовое поле PhoneNumber.

- в текстовое поле PhoneNumber теперь можно вставляться только разрешенные символы.
- в Contact.cs добавил константу MaxTextBoxSymbols для упрощения читаемости кода.

RockInHead committed:
style(View): исправил форматирование.

RockInHead committed:
refactoring(Contact): добавил новую константу.
